text/0088-fix-memory-limitiations-in-session-replays-access-pattern.md

```diff
@@ -0,0 +1,288 @@
+  - Feature is not present in a stable release.
+  - Feature is not present on any version of ClickHouse in Altinity.
+
+**Questions**
+
+### 8. Use an Alternative OLAP Database
+
+**Proposal**
+
+OLAP Databases such as Apache Pinot support upserts which appear to be a key requirement for the Replays product. An aggregation state schema can be defined for merging columns in real-time https://docs.pinot.apache.org/basics/data-import/upsert.
+
+**Drawbacks**
+
+- Changes to the data model would be necessary. We will not be able to aggregate array columns.
+- Pinot has ordering constraints.
+  - If segments arrive after their successor has already been ingested the aggregated state will not contain those rows.
+- There are no existing Apache Pinot installations within the Sentry org.
+- The Replays team is not large enough or experienced enough to manage a Pinot installation.
+  - This would require another team assuming the burden for us.
+  - Otherwise, additional budget would need to be allocated to the Replays team to hire outside experts.
+- We need to re-write our application logic for querying the datastore.
+- Migration pains.
+
+**Questions**
+
+### 9. Use an Alternative OLTP Database
+
+**Proposal**
+
+OLTP databases such as PostgreSQL and AlloyDB support updates which appear to be a key requirement for the Replays product. Our scale is small enough that a sharded PostgreSQL database could handle it. Read volume is low relative to write volume. We could optimize our database for this use case.
+
+**Drawbacks**
+
+- Race conditions will require single-threaded processing of like replay-id events.
+- Duplicate messages will necessitate ordering requirements.
+- Always a possibility for dropped and duplicated data regardless of safe guards.
+- AlloyDB is still in developer preview on Google Cloud.
+- We need to re-write our application logic for querying the datastore.
+- Migration pains.
+
+**Questions**
+
+- Colton Allen asks: Is it possible to write to tables partitioned by the hour? When the hour expires the table is indexed for fast read performance. Replays on the most recent hour partition would be unavailable while we are writing to it. Does PostgreSQL expose behavior that allows us to query over multiple partitions transparently? Is this even necessary given our scale? Currently processing 200 messages per second.
+
+### 10. Manually Manage An Aggregated Materialized View
+
+**Proposal**
+
+Use a cron job, which runs at some `interval` (e.g. 1 hour), that would select the finished replays from the last `interval`, aggregate them, and write them to a materialized view or some destination table. We then alter the index page to reflect two different dataset. The "live" dataset and the "video-on-demand" dataset. A "live" page would fetch replays from the last `interval`. A "video-on-demand" dataset would function similarly to the current replays index page however it would only contain data that has been aggregated by the cron job.
```

Comment:
I can't speak to the details of the cron job described here in this solution, but I would be in favour of a solution proposal that does not display "in progress" replays on the Index page by default -- as this leads to many short (and 'unfinished') replays for customers that have a high volume.

Reply:
> I would be in favour of a solution proposal that does not display "in progress" replays on the Index page by default

This is good to know.  This gives us a lot of flexibility to solve the problem.

From a product perspective I agree as well.  Some of our large customers have low utility index pages because of the number of new replays.